core/src/com/unciv/ui/utils/KeyCharAndCode.kt

```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    companion object {
+        /** Tests presence of a physical keyboard - static here as convenience shortcut only */
+        val keyboardAvailable = Gdx.input.isPeripheralAvailable(Input.Peripheral.HardwareKeyboard)
+
+        // Convenience shortcuts for frequently used constants
+        /** Android back, assigns ESC automatically as well */
+        val BACK = KeyCharAndCode(Input.Keys.BACK)
+        /** Automatically assigned for [BACK] */
+        val ESC = KeyCharAndCode(Input.Keys.ESCAPE)
+        /** Assigns [NUMPAD_ENTER] automatically as well */
+        val RETURN = KeyCharAndCode(Input.Keys.ENTER)
+        /** Automatically assigned for [RETURN] */
+        val NUMPAD_ENTER = KeyCharAndCode(Input.Keys.NUMPAD_ENTER)
+        val SPACE = KeyCharAndCode(Input.Keys.SPACE)
+        val DEL = KeyCharAndCode(Input.Keys.FORWARD_DEL)        // Gdx "DEL" is just plain wrong!
+        /** Guaranteed to be ignored by [KeyShortcutDispatcher] and never to be generated for an actual event, used as fallback to ensure no action is taken */
+        val UNKNOWN = KeyCharAndCode(Input.Keys.UNKNOWN)
+
+        // Kludges because we got crashes: java.lang.NoSuchMethodError: 'int kotlin.CharCodeKt.access$getCode$p(char)'
+        //TODO fixed by removing UncivServer from the desktop module - clean up comments and all uses
+        fun Char.toCode() = code
+//            try { code } catch (ex: Throwable) { null }
+//                ?: try { @Suppress("DEPRECATION") toInt() } catch (ex: Throwable) { null }
+//                ?: 0
+        fun Int.makeChar() = Char(this)
+//            try { Char(this) } catch (ex: Throwable) { null }
+//                ?: try { toChar() } catch (ex: Throwable) { null }
+//                ?: Char.MIN_VALUE
+
+        /** mini-factory for control codes - case insensitive */
+        fun ctrl(letter: Char) = KeyCharAndCode((letter.toCode() and 31).makeChar(),0)
+
+        /** mini-factory for control codes from keyCodes */
+        fun ctrlFromCode(keyCode: Int): KeyCharAndCode {
+            val name = Input.Keys.toString(keyCode)
+            if (name.length != 1 || !name[0].isLetter()) return KeyCharAndCode(Char.MIN_VALUE, keyCode)
+            return ctrl(name[0])
+        }
+
+        /** mini-factory for KeyCharAndCode values to be compared by character, not by code */
+        fun ascii(char: Char) = KeyCharAndCode(char.lowercaseChar(), 0)
+
+        /** factory maps a Char to a keyCode if possible, returns a Char-based instance otherwise */
+        fun mapChar(char: Char): KeyCharAndCode {
+            val code = Input.Keys.valueOf(char.uppercaseChar().toString())
+            return if (code == -1) KeyCharAndCode(char,0) else KeyCharAndCode(Char.MIN_VALUE, code)
+        }
+    }
+}
+
+
+data class KeyShortcut(val key: KeyCharAndCode, val priority: Int = 0)
+
+
+open class KeyShortcutDispatcher {
+    private var shortcuts: MutableList<Pair<KeyShortcut, () -> Unit>>? = null
```

Comment:
That was to conserve memory, so that attachment objects without shortcuts (in case `onActivation` really is used often) would take less memory. But OK, rewritten.